lib/java/com/google/android/material/bottomsheet/BottomSheetDialogFragment.java

```diff
@@ -31,4 +39,87 @@
   public Dialog onCreateDialog(Bundle savedInstanceState) {
     return new BottomSheetDialog(getContext(), getTheme());
   }
+
+  @Override
+  public void dismiss() {
+    Dialog dialog = getDialog();
+    if (dialog != null) {
+      View bottomSheet = dialog.findViewById(R.id.design_bottom_sheet);
+      if (bottomSheet != null) {
+        BottomSheetBehavior<View> behavior = BottomSheetBehavior.from(bottomSheet);
+
+        dismissWithAnimation(behavior);
+      } else {
+        super.dismiss();
+      }
+    } else {
+      super.dismiss();
+    }
+  }
+
+  @Override
+  public void dismissAllowingStateLoss() {
+    Dialog dialog = getDialog();
+    if (dialog != null) {
+      View bottomSheet = dialog.findViewById(R.id.design_bottom_sheet);
+      if (bottomSheet != null) {
+        BottomSheetBehavior<View> behavior = BottomSheetBehavior.from(bottomSheet);
+
+        dismissWithAnimation(behavior, true);
+      } else {
+        super.dismissAllowingStateLoss();
+      }
+    } else {
+      super.dismissAllowingStateLoss();
+    }
+  }
+
+  private void dismissWithAnimation(@NonNull BottomSheetBehavior behavior) {
+    dismissWithAnimation(behavior, false);
+  }
+
+  private void dismissWithAnimation(@NonNull BottomSheetBehavior behavior,
+                                    boolean allowingStateLoss) {
+    if (behavior.getState() == BottomSheetBehavior.STATE_HIDDEN) {
+      if (allowingStateLoss) {
+        super.dismissAllowingStateLoss();
+      } else {
+        super.dismiss();
+      }
+    } else {
+      BottomSheetBehavior.BottomSheetCallback dismissCallback =
+          getDismissBehaviorCallback(allowingStateLoss);
+
+      behavior.setBottomSheetCallback(dismissCallback);
+
+      behavior.setState(BottomSheetBehavior.STATE_HIDDEN);
+    }
+  }
+
+  private BottomSheetBehavior.BottomSheetCallback getDismissBehaviorCallback(
+      final boolean allowingStateLoss) {
+
+    final WeakReference<BottomSheetDialogFragment> dialogFragmentWeakReference =
+        new WeakReference<>(this);
```

Comment:
@cketcham After the rebase, I had to fix two compilation issues

- Had to make the `boolean allowingStateLoss` param in this function `final`, as it was accessed inside the callback.
- Had to change the generic for `WeakReference`
  - `WeakReference<AppCompatDialogFragment>` to `WeakReference<BottomSheetDialogFragment>`